New admin user list and user details pages

zarvox opened the pull request:
User list:

![user-list](https://cloud.githubusercontent.com/assets/307325/14924006/349e38aa-0df5-11e6-91bb-de06dce2c1e4.png)

Click on any particular user (or the "Manage" link for keyboard users), see details about them:

![user-details-self](https://cloud.githubusercontent.com/assets/307325/14924240/5a718b1c-0df6-11e6-8664-08a8d4ea7e83.png)

You cannot modify your own account, as shown above, but you can modify other users:

![user-details-avatar-multi-emails](https://cloud.githubusercontent.com/assets/307325/14924284/9656cc50-0df6-11e6-8b36-8623d0e91daa.png)

You can't demote org members to Visitor:

![cannot-demote-org-member](https://cloud.githubusercontent.com/assets/307325/14924263/7a61a736-0df6-11e6-8570-87e079402dfc.png)

Closes #1822.

neynah commented:
Looks great @zarvox. Can we left-align the column title "Properties"?

paulproteus commented:
Presumably we should rename 'guest' to 'visitor' in other places too before releasing this, to avoid confusion, including:

- Any error messages that the app provides

- In the docs

I'm OK with that being a separate pull request, but thought I should mention it for completeness. Since this is the code change that renames the concept from "Guest" to "Visitor", I'd prefer if @zarvox were the one to do those other fixes, but I'm also OK with him delegating that to a willing party. (I can be that willing party.)

paulproteus commented:
Oh, then again, this isn't live until we switch from "oldadmin" to "newadmin", right? So it's OK to delay that until then, I figure.

neynah commented:
Icons on the Users list should be top-aligned and the "Manage" link should also be styled as a button. Otherwise, looks great!

neynah commented:
On Users / Test it seems like only the verified e-mail is linked—is this intentional?

zarvox commented:
> Looks great @zarvox. Can we left-align the column title "Properties"?

Discussed in person, left-aligned both the column header and the contents for readability.

> (some things about guest vs visitor)

Yeah, let's do that when we do the switchover.

> Icons on the Users list should be top-aligned

Discussed in person, implemented.  Kind of a pain to get the paddings to line up nicely.

> and the "Manage" link should also be styled as a button. Otherwise, looks great!

Done.

I believe that addresses all the review comments given.

kentonv commented on shell/server/admin-server.js:
```diff
@@ -473,6 +473,90 @@ Meteor.publish("allUsers", function (token) {
+Meteor.publish("adminUserDetails", function (userId) {
+  if (!authorizedAsAdmin(undefined, this.userId)) return [];
+
+  // Reactive publish of any identities owned by
```
Finish sentence?

zarvox replied:
D'oh.

kentonv commented on shell/server/admin-server.js:
```diff
@@ -473,6 +473,90 @@ Meteor.publish("allUsers", function (token) {
+      },
+    });
+
+    identitySubs[identityId] = observeHandle;
```
If two accounts share an identity then I think this will leak the old handle.

kentonv replied:
I think you'll need some refcounting...

zarvox replied:
I don't think so - this publish only operates on a single account at a time.  So two accounts get two separate subscribe() calls, and two separate `identitySubs` instances, and two separate `observeHandle`s.

You might argue that I should be sharing the subscriptions between the publishes, and you'd probably be right.  I can rework this code to do so, but given that you're fairly unlikely to be looking at too many of these things at once in the first place, and it's only available to admins, I didn't think it was worth the complexity in the first pass.

kentonv replied:
OK, I think you should add some defensive code to detect a double-subscribe (and double-unsubscribe). If somehow someone ends up with the same identity listed twice on their account, it shouldn't cause the server to leak mongo subscriptions. Plus I could easily imagine this code changing later on to be used in a different way, and the name `refIdentity` sort of implies refcounting, so I'd hate for someone to assume that's what it does and have a silent leak...

kentonv replied:
To be clear, the code could just throw an exception... though then we have to ask whether the context is exception-safe. Where is RAII when you need it?

mitar replied:
> it shouldn't cause the server to leak mongo subscriptions

Duplicate subscriptions are de-duplicated by Meteor. I would not worry about such type of optimizations. It seems premature optimizations to me.

kentonv replied:
@mitar Avoiding memory leaks is not "optimization". I think you may misunderstand the problem here.

mitar replied:
Oh, oh. You are right.

You are manually doing related publish here? This is so error prone. Why not using some packages which are trying really hard to do this properly? Why reinventing the wheel? (Of course I have my [own](https://github.com/peerlibrary/meteor-reactive-publish), but not everyone will like my approach. I hear [this one](https://github.com/englue/meteor-publish-composite) is pretty popular as well.

kentonv commented on shell/server/admin-server.js:
```diff
@@ -473,6 +473,90 @@ Meteor.publish("allUsers", function (token) {
+  };
+
+  const refIdentity = (identityId) => {
+    const cursor = Meteor.users.find({ _id: identityId });
```
So we're subscribing to each identity separately? Will be interesting to see what happens on Oasis... I guess the only solution is pagination, though.

zarvox replied:
Note that this publish is only used on the user details page, so you're only calling this `subscribe()` on a single account ID.

zarvox replied:
For reference, the user-list page subscribes to the "allUsers" publish five lines above the start of this function.

kentonv replied:
Ohh, yeah, I should have read more carefully.

kentonv commented on shell/server/admin-server.js:
```diff
@@ -473,6 +473,90 @@ Meteor.publish("allUsers", function (token) {
+    });
+  });
+
+  this.ready();
```
I forget, does observe() call callbacks for the initial set of results synchronously? If so comment here as a reminder. If not then it's too early to call ready()...

zarvox replied:
Yes.  http://docs.meteor.com/#/full/observe says

> Before observe returns, added (or addedAt) will be called zero or more times to deliver the initial results of the query.

I'll add the comment.

kentonv commented:
"Make Admin" / "Make User" / "Make Visitor" seems like confusing wording to me... like I'm making a new one, or something. Maybe "Promote to Admin", "Promote to User", "Demote to Visitor"?

zarvox commented:
Stylecheck failure was introduced by #1923 which Garply never ran tests on, since Mitar isn't on the whitelist and no one flagged the patch as okay to test.  Added a fix to the end of this patch series.

mitar commented:
Whitelist me. ;-)

zarvox commented:
Added guards.

neynah commented:
@zarvox Not urgent but I forgot to mention "Show" should be inline with the search box. Also, some padding(10-15pxish) below the "Visitor" checkbox would reduce clutter.

Also, the search box takes up more space than it does in our current UI(although I don't believe this is currently standardized). Maybe we should chat at some point about how much horizontal space a search bar should take. I'm thinking 50-70% depending on the size of the screen.

zarvox commented:
I can be talked down from having the search bar take up the full available width, but in the absence of more precise specifications in the mockups, I will generally attempt to approximate a grid and assign any remaining space to the interactive elements or important presentational content, since larger click targets and readable text are generally preferable.

That said, some things benefit from having their width capped - readability is improved around 65 characters per line, and so forth.

If you think this sort of issue is likely to come up in the future, let's come up with a way to communicate this sort of intent more clearly in the mockups.  From my side, it's hard to infer from a static picture with no gridlines, size comments, or sub-boxes what is intended to happen at varying widths!
